### ISSUE TITLE
Feature request: `categorical_log_lpmf` and `categorical_log_rng`

### ISSUE THREAD

**Issue by mhollanders** [stan-dev/stan#3329]
Hi all,

I'm working on an N-mixture model and one way to get the posterior distribution for latent abundance is (I believe) to take a categorical draw of the probabilities associated with each abundance value; see code below:

```{stan}
generated quantities {
  vector[I] log_lik;
  array[I] int N;
  {
    vector[K] lp;
    for (i in 1:I) {
      for (k in max_y[i]:K) {
        lp[k] = poisson_lpmf(k | lambda[i]) + binomial_lpmf(y[i, 1:J[i]] | k, p[i, 1:J[i]]);
      }
      log_lik[i] = log_sum_exp(lp[max_y[i]:K]);
      N[i] = categorical_rng(exp(lp[max_y[i]:K] - log_lik[i])) + max_y[i] - 1;
    }
  }
}
```

The last line could be made more efficient if we didn't have to exponentiate the log probabilities in `categorical_rng` and we could instead write:

```{stan}
N[i] = categorical_log_rng(lp[max_y[i]:K] - log_lik[i]) + max_y[i] - 1;
```

Does it make sense to add this to the Stan language?

Thanks,

Matt

**Comment by nhuurre** [comment]
This [already exists](https://mc-stan.org/docs/functions-reference/bounded_discrete_distributions.html#categorical-distribution); it's called `categorical_logit_rng`.

**Comment by mhollanders** [comment]
> This [already exists](https://mc-stan.org/docs/functions-reference/bounded_discrete_distributions.html#categorical-distribution); it's called `categorical_logit_rng`.

That's for log odds, not log probabilities, right?

**Comment by nhuurre** [comment]
"odds" is just a ratio of probabilities. When you write `exp(lp[max_y[i]:K] - log_lik[i])` that is the softmax function in `categorical_logit`.

**Comment by mhollanders** [comment]
> "odds" is just a ratio of probabilities. When you write `exp(lp[max_y[i]:K] - log_lik[i])` that is the softmax function in `categorical_logit`.

Wow, I had no idea. Sorry about that!

**Comment by bob-carpenter** [comment]
That was my fault on the terminology.  The ML world calls them "logits" and I just failed to work through the algebra.  The inputs to softmax are (unnormalized) log probs, not log odds.  If you pin one of the input dimensions to 0 as is traditional, then they're log odds relative to the fixed value.  

I wrote up a [note about log vs. logit and softmax on Gelman's blog](https://statmodeling.stat.columbia.edu/2024/12/26/those-are-unnormalized-log-probabilities-not-logits-in-your-neural-networks-final-layer/).
